pkg/ovsdb/transact.go

```diff
@@ -0,0 +1,1094 @@
+	// XXX(alexey): can we just run on original ?
+	mutated := &map[string]interface{}{}
+	copier.Copy(mutated, original)
+	for _, mt := range mutations {
+		mutation, err := NewMutation(mt.([]interface{}))
+		if err != nil {
+			return err
+		}
+		err = mutation.Mutate(mutated)
+		if err != nil {
+			return err
+		}
+	}
+	copier.Copy(original, mutated)
+	return nil
+}
+
+func RowUpdate(original *map[string]interface{}, update map[string]interface{}) error {
+	for k, v := range update {
+		switch k {
+		case COL_UUID, COL_VERSION:
+			return errors.New(E_CONSTRAINT_VIOLATION)
+		}
+		(*original)[k] = v
+	}
+	return nil
+}
+
+func etcdGetTable(txn *Transaction, table string) {
+	prefix := makePrefix(txn.request.DBName, table)
+	etcdOp := clientv3.OpGet(prefix, clientv3.WithPrefix())
+	txn.etcdThen = append(txn.etcdThen, etcdOp)
+}
+
+func etcdGetRow(txn *Transaction, table, uuid string) {
+	key := makeKey(txn.request.DBName, table, uuid)
+	etcdOp := clientv3.OpGet(key)
+	txn.etcdThen = append(txn.etcdThen, etcdOp)
+}
+
+func etcdGetTableOrRow(txn *Transaction, table, uuid string) {
+	if uuid != "" {
+		etcdGetRow(txn, table, uuid)
+	} else {
+		etcdGetTable(txn, table)
+	}
+}
+
+func etcdGetByWhere(txn *Transaction, ovsOp *libovsdb.Operation, ovsResult *libovsdb.OperationResult) error {
+	uuid, err := doesWhereContainCondTypeUUID(ovsOp.Where)
+	if err != nil {
+		ovsResult.Error = err.Error()
+		return err
+	}
+	etcdGetTableOrRow(txn, ovsOp.Table, uuid)
+	return nil
+}
+
+func etcdPutRow(txn *Transaction, table, uuid string, row *map[string]interface{}) error {
+	key := makeKey(txn.request.DBName, table, uuid)
+	setRowUUID(row, uuid)
+	val, err := makeValue(row)
+	if err != nil {
+		return err
+	}
+	etcdOp := clientv3.OpPut(key, val)
+
+	/* remove any duplicate keys from prev operations */
+	newThen := []clientv3.Op{}
+	for _, op := range txn.etcdThen {
+		v := reflect.ValueOf(op)
+		f := v.FieldByName("key")
+		k := f.Bytes()
+		if string(k) != key {
+			newThen = append(newThen, op)
+		}
+	}
+	txn.etcdThen = newThen
+
+	txn.etcdThen = append(txn.etcdThen, etcdOp)
+	return nil
+}
+
+func etcdDelRow(txn *Transaction, table, uuid string) error {
+	key := makeKey(txn.request.DBName, table, uuid)
+	etcdOp := clientv3.OpDelete(key)
+	txn.etcdThen = append(txn.etcdThen, etcdOp)
+	return nil
+}
+
+/* insert */
+func preInsert(txn *Transaction, ovsOp *libovsdb.Operation, ovsResult *libovsdb.OperationResult) error {
+	if ovsOp.UUIDName == "" {
+		return nil
+	}
+	etcdGetTable(txn, ovsOp.Table)
+	return nil
+}
+
+func doInsert(txn *Transaction, ovsOp *libovsdb.Operation, ovsResult *libovsdb.OperationResult) error {
+	if ovsOp.UUIDName != "" {
+		for _, row := range txn.data.Table(ovsOp.Table) {
+			if (*row)[COL_UUIDNAME] == ovsOp.UUIDName {
+				err := errors.New(E_DUP_UUIDNAME)
+				ovsResult.Error = err.Error()
+				return err
+			}
+		}
+	}
+	ovsResult.Count = ovsResult.Count + 1
```

Comment:
it should be taken from the etcd response

Reply:
if we can ensure the database did not change then this result is accurate, if not then we need to use the value calculated from the etcd response (if we optimize delete then this may be relevant).